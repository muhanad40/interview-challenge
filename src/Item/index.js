import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Item extends Component {
  constructor() {
    super();
    this.onClick = this.onClick.bind(this);
  }

  onClick() {
    const {
      id,
      name,
      dietaries,
    } = this.props;

    this.props.onClick({ id, name, dietaries });
  }

  render() {
    return (
      <li className="item" onClick={this.onClick}>
        <h2>{this.props.name}</h2>
        <p>
          {this.props.dietaries.map(dietary => <span key={dietary} className="dietary">{dietary}</span>)}
        </p>
        {this.props.isRemovable &&
          <button onClick={this.props.onRemove} className="remove-item">x</button>
        }
      </li>
    );
  }
}

Item.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  dietaries: PropTypes.arrayOf(PropTypes.string).isRequired,
  isRemovable: PropTypes.bool,
  onRemove: PropTypes.func,
  onClick: PropTypes.func,
};

export default Item;
