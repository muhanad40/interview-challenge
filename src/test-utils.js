import React from 'react';
import { shallow } from 'enzyme';

export function createRenderer(Component, props) {
  return (extraProps) => shallow(<Component {...props} {...extraProps} />);
}
