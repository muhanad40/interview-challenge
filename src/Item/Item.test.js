require('../test-setup');

import Item from './index';
import { createRenderer } from '../test-utils';

describe('Item', () => {
  let render;
  const initProps = {
    id: 1234,
    name: 'Kale Caesar Pasta, Turmeric Satay Broccoli & Lemon Cashew Greens',
    dietaries: ['v', 've', 'df', 'gf', 'n!'],
  }

  beforeAll(() => {
    render = createRenderer(Item, initProps);
  });

  it('should render correctly', () => {
    expect(render()).toMatchSnapshot();
  });

  it('should render remove button if `isRemovable` prop is set', () => {
    expect(render({
      isRemovable: true,
    }).find('button')).toHaveLength(1);
  });

  it('should call remove callback when remove button is clicked', () => {
    const mockCallback = jest.fn();
    const item = render({
      isRemovable: true,
      onRemove: mockCallback,
    });

    item.find('button').simulate('click');

    expect(mockCallback).toHaveBeenCalledTimes(1);
  });

  it('should call `onClick` callback prop when clicked and pass item as argument', () => {
    const mockCallback = jest.fn();
    const item = render({
      onClick: mockCallback,
    });

    item.simulate('click');

    expect(mockCallback).toHaveBeenCalledTimes(1);
    expect(mockCallback).toHaveBeenCalledWith(initProps);
  });
});
