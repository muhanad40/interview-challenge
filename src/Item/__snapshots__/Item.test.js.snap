// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item should render correctly 1`] = `
<li
  className="item"
  onClick={[Function]}
>
  <h2>
    Kale Caesar Pasta, Turmeric Satay Broccoli & Lemon Cashew Greens
  </h2>
  <p>
    <span
      className="dietary"
      key="v"
    >
      v
    </span>
    <span
      className="dietary"
      key="ve"
    >
      ve
    </span>
    <span
      className="dietary"
      key="df"
    >
      df
    </span>
    <span
      className="dietary"
      key="gf"
    >
      gf
    </span>
    <span
      className="dietary"
      key="n!"
    >
      n!
    </span>
  </p>
</li>
`;
